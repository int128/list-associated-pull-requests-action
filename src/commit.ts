type Commit = {
  commitId: string
  committedDate: Date
}

export const findEarliestCommit = (commits: Commit[]): Commit => {
  let earliestCommit = commits[0]
  for (const commit of commits) {
    if (commit.committedDate.getTime() < earliestCommit.committedDate.getTime()) {
      earliestCommit = commit
    }
  }
  return earliestCommit
}
