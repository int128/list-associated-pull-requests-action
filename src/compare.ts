import * as core from '@actions/core'
import { GitHub } from '@actions/github/lib/utils'

type Octokit = InstanceType<typeof GitHub>

type Inputs = {
  owner: string
  repo: string
  base: string
  head: string
}

type Outputs = {
  commitIds: Set<string>
  earliestCommitId: string
  earliestCommitDate: Date
}

export const compareCommits = async (octokit: Octokit, inputs: Inputs): Promise<Outputs> => {
  const compareIterator = octokit.paginate.iterator(octokit.rest.repos.compareCommitsWithBasehead, {
    owner: inputs.owner,
    repo: inputs.repo,
    basehead: `${inputs.base}...${inputs.head}`,
    per_page: 100,
  })

  const commitIds = new Set<string>()
  let earliestCommitDate = new Date()
  let earliestCommitId = ''
  for await (const compare of compareIterator) {
    core.info(`Received ${compare.data.commits.length} commits of total ${compare.data.total_commits}`)
    core.info(`Remaining rate limit: ${compare.headers['x-ratelimit-remaining'] ?? '?'}`)

    for (const commit of compare.data.commits) {
      commitIds.add(commit.sha)
      if (commit.commit.committer?.date) {
        const d = new Date(commit.commit.committer.date)
        if (d < earliestCommitDate) {
          earliestCommitDate = d
          earliestCommitId = commit.sha
        }
      }
    }
  }
  return { commitIds, earliestCommitId, earliestCommitDate }
}
