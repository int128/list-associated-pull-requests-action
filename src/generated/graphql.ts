import * as Types from './graphql-types';

export type AssociatedPullRequestsInCommitHistoryOfSubTreeQueryVariables = Types.Exact<{
  owner: Types.Scalars['String'];
  name: Types.Scalars['String'];
  expression: Types.Scalars['String'];
  path: Types.Scalars['String'];
  since: Types.Scalars['GitTimestamp'];
  historySize: Types.Scalars['Int'];
  historyAfter?: Types.InputMaybe<Types.Scalars['String']>;
}>;


export type AssociatedPullRequestsInCommitHistoryOfSubTreeQuery = { __typename?: 'Query', rateLimit?: { __typename?: 'RateLimit', cost: number } | null, repository?: { __typename?: 'Repository', object?: { __typename: 'Blob' } | { __typename: 'Commit', history: { __typename?: 'CommitHistoryConnection', totalCount: number, pageInfo: { __typename?: 'PageInfo', hasNextPage: boolean, endCursor?: string | null }, nodes?: Array<{ __typename?: 'Commit', oid: string, associatedPullRequests?: { __typename?: 'PullRequestConnection', nodes?: Array<{ __typename?: 'PullRequest', number: number, author?: { __typename?: 'Bot', login: string } | { __typename?: 'EnterpriseUserAccount', login: string } | { __typename?: 'Mannequin', login: string } | { __typename?: 'Organization', login: string } | { __typename?: 'User', login: string } | null } | null> | null } | null } | null> | null } } | { __typename: 'Tag' } | { __typename: 'Tree' } | null } | null };
