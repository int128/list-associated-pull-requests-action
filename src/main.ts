import * as core from '@actions/core'
import { run } from './run'

const main = async (): Promise<void> => {
  const outputs = await run({
    token: core.getInput('token', { required: true }),
    pullRequest: parseInt(core.getInput('pull-request')) || undefined,
    base: core.getInput('base') || undefined,
    head: core.getInput('head') || undefined,
    groupByPaths: core.getMultilineInput('group-by-paths'),
    showOthersGroup: core.getBooleanInput('show-others-group', { required: true }),
  })

  core.setOutput('body', outputs.body)
}

main().catch((e) => core.setFailed(e instanceof Error ? e : String(e)))
