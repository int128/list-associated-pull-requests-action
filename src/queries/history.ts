import { GitHub } from '@actions/github/lib/utils'
import {
  AssociatedPullRequestsInCommitHistoryOfSubTreeQuery,
  AssociatedPullRequestsInCommitHistoryOfSubTreeQueryVariables,
} from '../generated/graphql'

type Octokit = InstanceType<typeof GitHub>

const query = /* GraphQL */ `
  query associatedPullRequestsInCommitHistoryOfSubTree(
    $owner: String!
    $name: String!
    $expression: String!
    $path: String!
    $since: GitTimestamp!
    $after: String
  ) {
    rateLimit {
      cost
    }
    repository(owner: $owner, name: $name) {
      object(expression: $expression) {
        __typename
        ... on Commit {
          history(path: $path, since: $since, first: 100, after: $after) {
            totalCount
            pageInfo {
              hasNextPage
              endCursor
            }
            nodes {
              oid
              associatedPullRequests(first: 1, orderBy: { field: CREATED_AT, direction: ASC }) {
                nodes {
                  number
                  author {
                    login
                  }
                }
              }
            }
          }
        }
      }
    }
  }
`

export const getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery = async (
  o: Octokit,
  v: AssociatedPullRequestsInCommitHistoryOfSubTreeQueryVariables
): Promise<AssociatedPullRequestsInCommitHistoryOfSubTreeQuery> => await o.graphql(query, v)
