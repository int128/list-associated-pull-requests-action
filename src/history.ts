import * as core from '@actions/core'
import { GitHub } from '@actions/github/lib/utils'
import { AssociatedPullRequestsInCommitHistoryOfSubTreeQuery } from './generated/graphql'
import { getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery } from './queries/history'

type Octokit = InstanceType<typeof GitHub>

type Inputs = {
  owner: string
  repo: string
  ref: string
  path: string
  since: Date
  sinceCommitId: string
}

export type Commit = {
  commitId: string
  pull?: {
    number: number
    author: string
  }
}

export const getCommitHistory = async (octokit: Octokit, inputs: Inputs): Promise<Commit[]> => {
  const commits: Commit[] = []
  for (let afterCursor: string | undefined; ; ) {
    const page = await getCommitHistoryByCursor(octokit, inputs, afterCursor)
    commits.push(...page.commits)
    if (page.hasNextPage === false) {
      break
    }
    afterCursor = page.endCursor
  }
  return commits
}

type CommitHistoryPage = {
  commits: Commit[]
  hasNextPage: boolean
  endCursor?: string
}

const getCommitHistoryByCursor = async (
  octokit: Octokit,
  inputs: Inputs,
  afterCursor?: string
): Promise<CommitHistoryPage> => {
  core.startGroup(
    `AssociatedPullRequestsInCommitHistoryOfSubTreeQuery (${inputs.ref}, ${inputs.path}, ${inputs.since.toISOString()})`
  )
  const q = await getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(octokit, {
    owner: inputs.owner,
    name: inputs.repo,
    expression: inputs.ref,
    path: inputs.path,
    since: inputs.since,
    after: afterCursor,
  })
  core.info(JSON.stringify(q, undefined, 2))
  core.endGroup()

  return parseAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(q, inputs.sinceCommitId)
}

export const parseAssociatedPullRequestsInCommitHistoryOfSubTreeQuery = (
  q: AssociatedPullRequestsInCommitHistoryOfSubTreeQuery,
  sinceCommitId: string
): CommitHistoryPage => {
  if (q.repository?.object?.__typename !== 'Commit') {
    throw new Error(`unexpected typename ${String(q.repository?.object?.__typename)} !== Commit`)
  }

  const nodes = filterNodes(q, sinceCommitId)
  core.startGroup(`Found ${nodes.length} commit(s)`)
  const commits: Commit[] = []
  for (const node of nodes) {
    if (!node.associatedPullRequests?.nodes?.length) {
      core.info(`${node.oid} -> no pull request`)
      commits.push({ commitId: node.oid })
      continue
    }
    for (const pull of node.associatedPullRequests.nodes) {
      if (pull?.number === undefined) {
        continue
      }
      core.info(`${node.oid} -> #${pull.number}`)
      commits.push({
        commitId: node.oid,
        pull: {
          number: pull.number,
          author: pull.author?.login ?? '',
        },
      })
    }
  }
  core.endGroup()

  return {
    commits,
    hasNextPage: q.repository.object.history.pageInfo.hasNextPage,
    endCursor: q.repository.object.history.pageInfo.endCursor ?? undefined,
  }
}

const filterNodes = (q: AssociatedPullRequestsInCommitHistoryOfSubTreeQuery, sinceCommitId: string) => {
  if (q.repository?.object?.__typename !== 'Commit') {
    throw new Error(`unexpected typename ${String(q.repository?.object?.__typename)} !== Commit`)
  }
  const nodes = []
  for (const node of q.repository.object.history.nodes ?? []) {
    if (node == null) {
      continue
    }
    nodes.push(node)
    if (node.oid === sinceCommitId) {
      break
    }
  }
  return nodes
}
