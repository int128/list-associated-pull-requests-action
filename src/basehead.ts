import * as core from '@actions/core'
import * as github from '@actions/github'
import { ChangeSet, findChangeSet } from './history'
import { getCommit } from './queries/commit'
import { getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery } from './queries/history'

type Inputs = {
  token: string
  base: string
  head: string
  path: string
  groupBySubPaths: string[]
}

type Outputs = {
  body: string
  associatedPullRequests: number[]
}

export const computeChangeSetBetweenBaseHead = async (inputs: Inputs): Promise<Outputs> => {
  const octokit = github.getOctokit(inputs.token)

  const baseCommit = await getCommit(octokit, {
    owner: github.context.repo.owner,
    name: github.context.repo.repo,
    expression: inputs.base,
  })
  core.startGroup(`Base commit of ${inputs.base}`)
  core.info(JSON.stringify(baseCommit, undefined, 2))
  core.endGroup()
  if (baseCommit.repository?.object?.__typename !== 'Commit') {
    throw new Error(`unexpected typename ${String(baseCommit.repository?.object?.__typename)} !== Commit`)
  }

  const history = await getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(octokit, {
    owner: github.context.repo.owner,
    name: github.context.repo.repo,
    expression: inputs.head,
    path: inputs.path,
    since: baseCommit.repository.object.committedDate,
  })
  core.startGroup(`Commit history on ${inputs.head} since ${baseCommit.repository.object.committedDate}`)
  core.info(JSON.stringify(history, undefined, 2))
  core.endGroup()
  const changeSet = findChangeSet(history, baseCommit.repository.object.oid)

  if (inputs.groupBySubPaths.length === 0) {
    return {
      body: [...changeSet.pullOrCommits].map((s) => `- ${s}`).join('\n'),
      associatedPullRequests: [...changeSet.pulls],
    }
  }

  const subPathChangeSets = new Map<string, ChangeSet>()
  for (const subPath of inputs.groupBySubPaths) {
    const history = await getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(octokit, {
      owner: github.context.repo.owner,
      name: github.context.repo.repo,
      expression: inputs.head,
      path: subPath,
      since: baseCommit.repository.object.committedDate,
    })
    core.startGroup(`Commit history of subpath ${subPath}`)
    core.info(JSON.stringify(history, undefined, 2))
    core.endGroup()
    const changeSet = findChangeSet(history, baseCommit.repository.object.oid)
    subPathChangeSets.set(subPath, changeSet)
  }

  const otherPullOrCommits = new Set(changeSet.pullOrCommits)
  const body = []
  for (const [subPath, { pullOrCommits }] of subPathChangeSets) {
    if (pullOrCommits.size > 0) {
      body.push(`### ${subPath}`)
      for (const pullOrCommit of pullOrCommits) {
        body.push(`- ${pullOrCommit}`)
        otherPullOrCommits.delete(pullOrCommit)
      }
    }
  }
  if (otherPullOrCommits.size > 0) {
    body.push(`### Others`)
    for (const pullOrCommit of otherPullOrCommits) {
      body.push(`- ${pullOrCommit}`)
    }
  }

  return {
    body: body.join('\n'),
    associatedPullRequests: [...changeSet.pulls],
  }
}
