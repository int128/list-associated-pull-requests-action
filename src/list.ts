import * as core from '@actions/core'
import { GitHub } from '@actions/github/lib/utils'
import { compareCommits } from './compare'
import { Commit, getCommitHistory } from './history'

type Octokit = InstanceType<typeof GitHub>

type Inputs = {
  owner: string
  repo: string
  base: string
  head: string
  groupByPaths: string[]
  showOthersGroup: boolean
}

type Outputs = {
  commitsByPath: Map<string, Commit[]>
  others?: Commit[]
}

export const listAssociatedPullRequests = async (octokit: Octokit, inputs: Inputs): Promise<Outputs> => {
  core.info(`Compare ${inputs.base} and ${inputs.head}`)
  const compare = await compareCommits(octokit, {
    owner: inputs.owner,
    repo: inputs.repo,
    base: inputs.base,
    head: inputs.head,
  })
  core.info(`commits = ${compare.commitIds.size}`)
  core.info(`earliestCommit = ${compare.earliestCommitId} (${compare.earliestCommitDate.toISOString()})`)

  const commitsByPath = new Map<string, Commit[]>()
  for (const path of inputs.groupByPaths) {
    const commitHistory = await getCommitHistory(octokit, {
      owner: inputs.owner,
      repo: inputs.repo,
      ref: inputs.head,
      path,
      since: compare.earliestCommitDate,
      sinceCommitId: compare.earliestCommitId,
    })
    const commits = commitHistory.filter((commit) => compare.commitIds.has(commit.commitId))
    commitsByPath.set(path, commits)
  }
  if (!inputs.showOthersGroup) {
    return { commitsByPath }
  }

  const rootCommitHistory = await getCommitHistory(octokit, {
    owner: inputs.owner,
    repo: inputs.repo,
    ref: inputs.head,
    path: '.',
    since: compare.earliestCommitDate,
    sinceCommitId: compare.earliestCommitId,
  })
  const rootCommits = rootCommitHistory.filter((commit) => compare.commitIds.has(commit.commitId))

  const otherCommits = new Map<string, Commit>()
  for (const commit of rootCommits) {
    otherCommits.set(commit.commitId, commit)
  }
  for (const [, commits] of commitsByPath) {
    for (const commit of commits) {
      otherCommits.delete(commit.commitId)
    }
  }

  return {
    commitsByPath,
    others: [...otherCommits.values()],
  }
}
