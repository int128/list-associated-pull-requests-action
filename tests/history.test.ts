import { AssociatedPullRequestsInCommitHistoryOfSubTreeQuery } from '../src/generated/graphql'
import { parseAssociatedPullRequestsInCommitHistoryOfSubTreeQuery } from '../src/history'

/*
Commits of pull request #98.
You can get this from https://docs.github.com/en/graphql/overview/explorer with:

{
  "owner": "int128",
  "name": "list-associated-pull-requests-action",
  "path": ".",
  "expression": "test-fixture-head",
  "since": "2022-05-13T22:33:57Z"
}
*/
const fixtureQuery: AssociatedPullRequestsInCommitHistoryOfSubTreeQuery = {
  rateLimit: {
    cost: 1,
  },
  repository: {
    object: {
      __typename: 'Commit',
      history: {
        totalCount: 27,
        nodes: [
          {
            oid: '0f9bbfade221f8b590163c664e27c3839fffd3c8',
            associatedPullRequests: {
              nodes: [
                {
                  number: 97,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: 'f2ac78bee33a1a55186732a1f2fb6b7d68008a53',
            associatedPullRequests: {
              nodes: [
                {
                  number: 90,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: 'dcff886579270638a2146cf8ba5380f2240774c7',
            associatedPullRequests: {
              nodes: [
                {
                  number: 96,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: '4e216f67567bdbb723ad28b1d946c63b6cb09a8d',
            associatedPullRequests: {
              nodes: [
                {
                  number: 91,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: 'aba1fb9e861f34176727a527a6e5af8ba74f628c',
            associatedPullRequests: {
              nodes: [
                {
                  number: 89,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: 'ca85088adb0f411ef934833e2b4c47941794ac58',
            associatedPullRequests: {
              nodes: [
                {
                  number: 88,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: 'ebb87b274ad05f7e5d9a38c677f2e399041dae01',
            associatedPullRequests: {
              nodes: [
                {
                  number: 87,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: 'fc7d55b8f3e86b9f2b71c30d9e8578dffc40c807',
            associatedPullRequests: {
              nodes: [
                {
                  number: 86,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: '8e07a09848e1158d3326ec2ab9006077c087bd33',
            associatedPullRequests: {
              nodes: [
                {
                  number: 85,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: 'aed841c2c1550f2e2bac160837085856b6cc5120',
            associatedPullRequests: {
              nodes: [
                {
                  number: 84,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: '13868898d92652c27d889259bd11215f7a4551c1',
            associatedPullRequests: {
              nodes: [
                {
                  number: 83,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: 'd486e046fca314b3a42693a5ca878e47c69a89ed',
            associatedPullRequests: {
              nodes: [
                {
                  number: 82,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: '6df54289158cbd5a61c8604cf97b8a0f7d1847f7',
            associatedPullRequests: {
              nodes: [
                {
                  number: 81,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: '49a339c426a28b6e5abd3203fadcc749560b2362',
            associatedPullRequests: {
              nodes: [
                {
                  number: 80,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: 'd5414767c25010a8d42f44121dba0ee870c1da53',
            associatedPullRequests: {
              nodes: [
                {
                  number: 79,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: 'e1d47057c2a2eafd8ebb570c8f7a4a84dffb234e',
            associatedPullRequests: {
              nodes: [
                {
                  number: 78,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: '15a4eb0b903195d92d089454ab9ffa1c7598d3ef',
            associatedPullRequests: {
              nodes: [
                {
                  number: 77,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: '797d211a801b403772b703ed29e60dc148d708b8',
            associatedPullRequests: {
              nodes: [
                {
                  number: 76,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: 'b256f3b36b635040e4ce150eec4c4d2fdab3495c',
            associatedPullRequests: {
              nodes: [
                {
                  number: 75,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: '1a74c41db6587e71a6c97a2a64741d6152a88638',
            associatedPullRequests: {
              nodes: [
                {
                  number: 74,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: '6c696b80a95830b40ae8ed882bbb9bbe1972b24c',
            associatedPullRequests: {
              nodes: [
                {
                  number: 73,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: 'ead2316173fbd364d6d241b8c05df4a349c7821f',
            associatedPullRequests: {
              nodes: [
                {
                  number: 70,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: '4449d60e7f1ce5fd3d26750c66b5d708cbac1d20',
            associatedPullRequests: {
              nodes: [
                {
                  number: 71,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: 'f656f6812207d5301492cab12f399463e35fbdbb',
            associatedPullRequests: {
              nodes: [
                {
                  number: 69,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: '0ad635a9b2be184a6a02253bcdb090dab6e195c9',
            associatedPullRequests: {
              nodes: [
                {
                  number: 68,
                  author: {
                    login: 'int128',
                  },
                },
              ],
            },
          },
          {
            oid: '3d9cb2f7a130e4427676df5f6da9c06cd228d7d7',
            associatedPullRequests: {
              nodes: [
                {
                  number: 67,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
          {
            oid: '639b198eeec3ff282b00e53e936929b72c004817',
            associatedPullRequests: {
              nodes: [
                {
                  number: 64,
                  author: {
                    login: 'renovate',
                  },
                },
              ],
            },
          },
        ],
      },
    },
  },
}

test('parseAssociatedPullRequestsInCommitHistoryOfSubTreeQuery', () => {
  const sinceCommitId = '3d9cb2f7a130e4427676df5f6da9c06cd228d7d7'
  const commits = parseAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(fixtureQuery, sinceCommitId)
  expect(commits).toMatchSnapshot()
})
