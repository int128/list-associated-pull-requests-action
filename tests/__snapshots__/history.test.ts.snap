// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseAssociatedPullRequestsInCommitHistoryOfSubTreeQuery 1`] = `
{
  "commits": [
    {
      "commitId": "0f9bbfade221f8b590163c664e27c3839fffd3c8",
      "pull": {
        "author": "renovate",
        "number": 97,
      },
    },
    {
      "commitId": "f2ac78bee33a1a55186732a1f2fb6b7d68008a53",
      "pull": {
        "author": "renovate",
        "number": 90,
      },
    },
    {
      "commitId": "dcff886579270638a2146cf8ba5380f2240774c7",
      "pull": {
        "author": "int128",
        "number": 96,
      },
    },
    {
      "commitId": "4e216f67567bdbb723ad28b1d946c63b6cb09a8d",
      "pull": {
        "author": "int128",
        "number": 91,
      },
    },
    {
      "commitId": "aba1fb9e861f34176727a527a6e5af8ba74f628c",
      "pull": {
        "author": "int128",
        "number": 89,
      },
    },
    {
      "commitId": "ca85088adb0f411ef934833e2b4c47941794ac58",
      "pull": {
        "author": "int128",
        "number": 88,
      },
    },
    {
      "commitId": "ebb87b274ad05f7e5d9a38c677f2e399041dae01",
      "pull": {
        "author": "int128",
        "number": 87,
      },
    },
    {
      "commitId": "fc7d55b8f3e86b9f2b71c30d9e8578dffc40c807",
      "pull": {
        "author": "renovate",
        "number": 86,
      },
    },
    {
      "commitId": "8e07a09848e1158d3326ec2ab9006077c087bd33",
      "pull": {
        "author": "renovate",
        "number": 85,
      },
    },
    {
      "commitId": "aed841c2c1550f2e2bac160837085856b6cc5120",
      "pull": {
        "author": "renovate",
        "number": 84,
      },
    },
    {
      "commitId": "13868898d92652c27d889259bd11215f7a4551c1",
      "pull": {
        "author": "renovate",
        "number": 83,
      },
    },
    {
      "commitId": "d486e046fca314b3a42693a5ca878e47c69a89ed",
      "pull": {
        "author": "renovate",
        "number": 82,
      },
    },
    {
      "commitId": "6df54289158cbd5a61c8604cf97b8a0f7d1847f7",
      "pull": {
        "author": "renovate",
        "number": 81,
      },
    },
    {
      "commitId": "49a339c426a28b6e5abd3203fadcc749560b2362",
      "pull": {
        "author": "renovate",
        "number": 80,
      },
    },
    {
      "commitId": "d5414767c25010a8d42f44121dba0ee870c1da53",
      "pull": {
        "author": "renovate",
        "number": 79,
      },
    },
    {
      "commitId": "e1d47057c2a2eafd8ebb570c8f7a4a84dffb234e",
      "pull": {
        "author": "int128",
        "number": 78,
      },
    },
    {
      "commitId": "15a4eb0b903195d92d089454ab9ffa1c7598d3ef",
      "pull": {
        "author": "int128",
        "number": 77,
      },
    },
    {
      "commitId": "797d211a801b403772b703ed29e60dc148d708b8",
      "pull": {
        "author": "int128",
        "number": 76,
      },
    },
    {
      "commitId": "b256f3b36b635040e4ce150eec4c4d2fdab3495c",
      "pull": {
        "author": "int128",
        "number": 75,
      },
    },
    {
      "commitId": "1a74c41db6587e71a6c97a2a64741d6152a88638",
      "pull": {
        "author": "renovate",
        "number": 74,
      },
    },
    {
      "commitId": "6c696b80a95830b40ae8ed882bbb9bbe1972b24c",
      "pull": {
        "author": "int128",
        "number": 73,
      },
    },
    {
      "commitId": "ead2316173fbd364d6d241b8c05df4a349c7821f",
      "pull": {
        "author": "int128",
        "number": 70,
      },
    },
    {
      "commitId": "4449d60e7f1ce5fd3d26750c66b5d708cbac1d20",
      "pull": {
        "author": "renovate",
        "number": 71,
      },
    },
    {
      "commitId": "f656f6812207d5301492cab12f399463e35fbdbb",
      "pull": {
        "author": "int128",
        "number": 69,
      },
    },
    {
      "commitId": "0ad635a9b2be184a6a02253bcdb090dab6e195c9",
      "pull": {
        "author": "int128",
        "number": 68,
      },
    },
    {
      "commitId": "3d9cb2f7a130e4427676df5f6da9c06cd228d7d7",
      "pull": {
        "author": "renovate",
        "number": 67,
      },
    },
  ],
  "endCursor": undefined,
  "hasNextPage": false,
}
`;
